### ISSUE TITLE
How to write cython code  of struct in struct in the type declaration. 

### ISSUE THREAD

**Issue by JKVISION1101** [OpenMS/OpenMS#7287]
Hi, 

I am now writing FLASHDeconvAlgorithm.pxd file for wrapping run method.

the signature of run method is this
`void run(MSExperiment& map, std::vector<DeconvolvedSpectrum>& deconvolved_spectra, std::vector<FLASHDeconvHelperStructs::MassFeature>& deconvolved_feature);`

In this code, I don't know how to deal with `std::vector<FLASHDeconvHelperStructs::MassFeature>` to the cython code. 

First, I tried ` libcpp_vector[FLASHDeconvHelperStructs::MassFeature] & massfeature`

However, error occured. (**In cython, :: this scope resolution operator is working?**)

Second, I referenced TargetedExperimentHelper.pxd file

`ctypedef enum RTUnit "OpenMS::TargetedExperimentHelper::RetentionTime::RTUnit":`
 `           # wrap-attach:`
 `           #   RetentionTime`
  `          SECOND,        # RT stored in seconds`
  `          MINUTE,        # RT stored in minutes`
 `           UNKNOWN,       # no stored annotation`
  `          SIZE_OF_RTUNIT`

this pxd file used type definition(ctypedef) to write RTUnit instead of writing `OpenMS::TargetedExperimentHelper::RetentionTime::RTUnit`

So I changed my code

In the `FLASHDeconvHelperStruct.pxd` file, I inserted this code

`cdef extern from "<OpenMS/ANALYSIS/TOPDOWN/FLASHDeconvHelperStructs.h>" namespace "OpenMS::FLASHDeconvHelperStructs::MassFeature":`
 `   ctypedef struct MassFeature "OpenMS::FLASHDeconvHelperStructs::MassFeature"`

and In the `FLASHDeconvAlgorithm.pxd`
I writed `libcpp_vector[MassFeature] & massfeature` instead of `libcpp_vector[FLASHDeconvHelperStructs::MassFeature] & massfeature`

However, It also make error like this. ( It said that Name error, no converter for MassFeature in default dict)
![error](https://github.com/OpenMS/OpenMS/assets/120628409/0a0bcf89-e8df-4f81-8cd2-13afa09360e2)

In short, In the cython grammar, How can i deal with scope resolution operator which is needed for struct in struct .

**Comment by jpfeuffer** [comment]
You cannot use the namespace operator in anything interfacing to python (i.e. function or class declarations).
Only things like "extern .. from namespace" or maybe ctypedef.

Just wrap every struct X FlashDeconvHelperStructs as FD_X and then use FD_X in your other functions.

**Comment by jpfeuffer** [comment]
See MassTrace.pxd where we wrap Mass trace from Kernel with a different name because there is another MassTrace somewhere else in another C++ namespace.

https://github.com/OpenMS/OpenMS/blob/60e2905a7571d7b6a479dcd69141dfd2ea220b66/src/pyOpenMS/pxds/MassTrace.pxd#L9

**Comment by JKVISION1101** [comment]
@jpfeuffer 


In MassTrace case, 

`cdef extern from "<OpenMS/KERNEL/MassTrace.h>" namespace "OpenMS":`

   ` cdef cppclass Kernel_MassTrace "OpenMS::MassTrace": `

Class MassTrace name is replaced with Kernel_MassTrace.




In my case,

`cdef extern from "<OpenMS/ANALYSIS/TOPDOWN/FLASHDeconvHelperStructs.h>" namespace "OpenMS":`

`cdef cppclass MassFeature_FDHS "OpenMS::FLASHDeconvHelperStructs::MassFeature"`

I wrote code above. However, i am not sure that this code which i wrote above is right. could you check it?

Pyopenms build succeeded. 

Now, FLASHDeconvAlgorithm. run method was wrapped( it needed to be tested.)



Your advice has been instrumental in advancing  project.

Thank you very much.
